_posts/2018-08-03-automatic-bug-reporting.md

```diff
@@ -0,0 +1,111 @@
+good policies. To abide by the GDPR we cannot collect any Personally Identifying Information. 
+What is PII for a bug report? Most of the choices were very straightforward, but a few required
+careful work to preserve our desired behavior. Beyond avoiding sending file paths from a user’s
+machine or data from their code, we also cannot capture their code in our stack trace. If a 
+Visual Studio For Mac extension contains code that is not ours, we should not collect information
+about their crashes. It is simple enough to only report the UUID, token, and managed IL offset 
+(CIL metadata about the class) that all work together like a unique hash. Without the C# assembly
+file in question, you don’t know which hash goes to which file. It’s a primary key, but the
+people who have the mapping already have the private data. 
+
+We needed to balance that desire against the desire for a crash from two different versions of mono
+to look very similar or identical to the backend (Part 3). Mono already has unique hashing functions
+for metadata objects; we use it inside of the AOT compiler and runtime. We can then generate a 
+hash that is identical for two identical stack traces, while uploading a version-dependent unique
+reference with the main stack trace. What this means is that if two crashes have the same hash but 
+different information otherwise, they are the same crash. This preserves privacy while letting you 
+count how often each bug is hit. 
+
+To make this concrete, this is an example of what we send for a managed frame:
+
+```
+    {
+    "is_managed" : "true",
+    "guid" : "0845998F-6B70-4AA8-9214-6731378926A0",
+    "token" : "0x6003817",
+    "native_offset" : "0x1fd",
+    "il_offset" : "0x00071"
+    }
+```
+
+### Bigger Impacts ###
+
+This crash dump creation also represented a change in our philosophy with respect to error reporting.
+Rather than expecting the bug report submission to be the beginning of an ongoing conversation,
+we are now expecting it to be an anonymous message. This motivates our embedding information into the 
+dump that we would typically ask someone to collect for us by exporting an environment variable or
+attaching with a debugger. The dump needs to contain enough system clues without describing the 
+embedding application too closely and being creepy. 
+
+It is a much harder problem for a runtime than for a web server because we expose a lot of the 
+details of the underlying platform to a developer who chooses to poke and prod. The CPU model isn’t 
+likely to cause problems with a web server, but leads to  
+[impossible bugs](https://www.mono-project.com/news/2016/09/12/arm64-icache/) with mono.
+We expose these abstract state machines in the API that only roughly correlate to the bits in the machine registers.
+This train of thought with respect to state machines and logging lead to the flight recorders, which will be mentioned in a subsequent blog post. 
```

Comment:
It's hard to make the connection of state machines, logging and flight recorders.
Maybe explain that the concrete way we realize those state machines is through logging and flight recorders. Plus give some high level overview (a few words) of what a flight recorder is. IE:

`Flight recorder is the name we give to lightweight logging doing using a circular buffer`  - or what you feel it's appropriate to mention here.